### PR TITLE
mds: Reorganize structure and class members in mdstypes header

### PR DESCRIPTION
**Pull request by varshar16**
Fixes: https://tracker.ceph.com/issues/43426
Signed-off-by: Varsha Rao <varao@redhat.com>

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by varshar16** [comment]
Rebased

**Comment by batrick** [comment]
needs rebased again

**Comment by varshar16** [comment]
@batrick rebased

File: src/mds/mdstypes.h [hunk 1]
```diff
@@ -1456,12 +1443,15 @@ struct cap_reconnect_t {
 
   void dump(Formatter *f) const;
   static void generate_test_instances(std::list<cap_reconnect_t*>& ls);
+
+  string path;
+  mutable ceph_mds_cap_reconnect capinfo;
```
Comment (batrick) [inline src/mds/mdstypes.h]: `= {}`

File: src/mds/mdstypes.h [hunk 2]
```diff
@@ -1424,18 +1418,11 @@ namespace std {
...
 
-
 // cap info for client reconnect
 struct cap_reconnect_t {
-  string path;
-  mutable ceph_mds_cap_reconnect capinfo;
-  snapid_t snap_follows;
-  bufferlist flockbl;
-
   cap_reconnect_t() {
```
Comment (batrick) [inline src/mds/mdstypes.h]: drop constructor after chaning `capinfo` init below

File: src/mds/mdstypes.h [hunk 3]
```diff
@@ -1478,6 +1468,8 @@ struct snaprealm_reconnect_t {
 
   void dump(Formatter *f) const;
   static void generate_test_instances(std::list<snaprealm_reconnect_t*>& ls);
+
+  mutable ceph_mds_snaprealm_reconnect realm;
```
Comment (batrick) [inline src/mds/mdstypes.h]: `= {}`

File: src/mds/mdstypes.h [hunk 4]
```diff
@@ -1456,12 +1443,15 @@ struct cap_reconnect_t {
...
+  mutable ceph_mds_cap_reconnect capinfo;
+  snapid_t snap_follows = 0;
+  bufferlist flockbl;
 };
 WRITE_CLASS_ENCODER(cap_reconnect_t)
 
 struct snaprealm_reconnect_t {
-  mutable ceph_mds_snaprealm_reconnect realm;
-
   snaprealm_reconnect_t() {
```
Comment (batrick) [inline src/mds/mdstypes.h]: drop after fixing init below

File: src/mds/mdstypes.h [hunk 5]
```diff
@@ -1527,17 +1516,14 @@ struct old_cap_reconnect_t {
...
 
-
-// ================================================================
 // dir frag
-
 struct dirfrag_t {
-  inodeno_t ino = 0;
-  frag_t    frag;
-
   dirfrag_t() {}
```
Comment (batrick) [inline src/mds/mdstypes.h]: ditto

File: src/mds/mdstypes.h [hunk 6]
```diff
@@ -1579,17 +1567,6 @@ namespace std {
...
-
-
-// ================================================================
-
-#define META_POP_IRD     0
-#define META_POP_IWR     1
-#define META_POP_READDIR 2
-#define META_POP_FETCH   3
-#define META_POP_STORE   4
-#define META_NPOP        5
```
Comment (batrick) [inline src/mds/mdstypes.h]: These definitions can stay here.

File: src/mds/mdstypes.h [hunk 7]
```diff
@@ -1771,18 +1743,15 @@ class load_spread_t {
...
-  int last[MAX];
-  int p = 0, n = 0;
-  DecayCounter count;
 
-public:
-  load_spread_t() = delete;
   load_spread_t(const DecayRate &rate) : count(rate)
   {
     for (int i=0; i<MAX; i++)
       last[i] = -1;
```
Comment (batrick) [inline src/mds/mdstypes.h]: drop this loop and fix init below:

File: src/mds/mdstypes.h [hunk 8]
```diff
@@ -1800,35 +1769,28 @@ class load_spread_t {
   double get() const {
     return count.get();
   }
-};
-
 
+  int last[MAX];
```
Comment (batrick) [inline src/mds/mdstypes.h]: `= {-1, -1, -1, -1}`

Comment (batrick) [reply src/mds/mdstypes.h]: We can also use this opportunity to fix the type: `std::array<int, MAX>`

File: src/mds/mdstypes.h [hunk 9]
```diff
@@ -1800,35 +1769,28 @@ class load_spread_t {
...
 typedef std::pair<mds_rank_t, mds_rank_t> mds_authority_t;
 
-// -- authority delegation --
-// directory authority types
-//  >= 0 is the auth mds
-#define CDIR_AUTH_PARENT   mds_rank_t(-1)   // default
-#define CDIR_AUTH_UNKNOWN  mds_rank_t(-2)
-#define CDIR_AUTH_DEFAULT  mds_authority_t(CDIR_AUTH_PARENT, CDIR_AUTH_UNKNOWN)
-#define CDIR_AUTH_UNDEF    mds_authority_t(CDIR_AUTH_UNKNOWN, CDIR_AUTH_UNKNOWN)
-//#define CDIR_AUTH_ROOTINODE pair<int,int>( 0, -2)
```
Comment (batrick) [inline src/mds/mdstypes.h]: These can also stay

File: src/mds/mdstypes.h [hunk 10]
```diff
@@ -1800,35 +1769,28 @@ class load_spread_t {
...
-
 class MDSCacheObjectInfo {
 public:
-  inodeno_t ino = 0;
-  dirfrag_t dirfrag;
-  string dname;
-  snapid_t snapid;
-
   MDSCacheObjectInfo() {}
```
Comment (batrick) [inline src/mds/mdstypes.h]: drop

File: src/mds/mdstypes.h [hunk 11]
```diff
@@ -1121,9 +1118,6 @@ inline std::ostream& operator<<(std::ostream& out, const feature_bitset_t& s) {
  * metric_spec_t
  */
```
[the remark this answers is not in the document]
Comment (varshar16) [reply src/mds/mdstypes.h]: Sorry missed it.

File: src/mds/mdstypes.h [hunk 12]
```diff
@@ -80,20 +96,13 @@ constexpr fs_cluster_id_t FS_CLUSTER_ID_NONE = -1;
...
+public:
   mds_role_t(fs_cluster_id_t fscid_, mds_rank_t rank_)
     : fscid(fscid_), rank(rank_)
   {}
-  mds_role_t()
-    : fscid(FS_CLUSTER_ID_NONE), rank(MDS_RANK_NONE)
-  {}
-  bool operator<(mds_role_t const &rhs) const
-  {
+  mds_role_t() {}
```
[the remark this answers is not in the document]
Comment (varshar16) [reply src/mds/mdstypes.h]: Can't be dropped as no default constructor is created.

**Comment by varshar16** [comment]
Not dropping few of the constructors as no default constructor is created.

**Comment by varshar16** [comment]
jenkins retest this please

**Pull request merged**